src/toast/tod/crosstalk.py

```diff
@@ -0,0 +1,587 @@
+I.e. in order to provide a block-diagonal crosstalk matrix,
+each block (together with the names of the detectors in that block)
+should write to a seperate HDF5 file to avoid unnessary computation.
+"""
+
+import argparse
+from pathlib import Path
+from typing import TYPE_CHECKING
+
+import h5py
+import numpy as np
+
+try:
+    from numba import jit
+except ImportError:
+    jit = None
+
+from ..mpi import get_world
+from ..op import Operator
+from ..utils import Logger
+
+if TYPE_CHECKING:
+    from typing import List, Union  # Optional
+
+    import toast
+
+# py37+
+# world_comm: Optional[toast.mpi.Comm]
+# PROCS: int
+# RANK: int
+world_comm, PROCS, RANK = get_world()
+logger = Logger.get()
+IS_SERIAL = PROCS == 1
+
+
+def _fma(
+    out: 'np.ndarray[np.float64]',
+    weights: 'np.ndarray[np.float64]',
+    *arrays: 'np.ndarray[np.float64]',
+):
+    """Simple fused multiply–add, compiled to avoid Python memory implications.
+
+    :param out: must be zero array in the same shape of each in `arrays`
+
+    If not compiled, a lot of Python objects will be created,
+    and as the Python garbage collector is inefficient,
+    it would have larger memory footprints.
+    """
+    for weight, array in zip(weights, arrays):
+        out += weight * array
+
+
+if jit is None:
+    logger.warning(
+        'Numba not present. '
+        '_fma in crosstalk will have more intermediate Numpy array objects '
+        'created that uses more memory.'
+    )
+else:
+    # cache is False to avoid IO on HPC.
+    _fma = jit(_fma, nopython=True, nogil=True, parallel=True, cache=False)
+
+
+def add_crosstalk_args(parser: 'argparse.ArgumentParser'):
+    """Add crosstalk args to argparse.ArgumentParser object."""
+    parser.add_argument(
+        "--crosstalk-matrix",
+        type=Path,
+        nargs='*',
+        required=False,
+        help="input path(s) to crosstalk matrix in HDF5 container.",
+    )
+    parser.add_argument(
+        "--crosstalk-matrix-debug",
+        type=bool,
+        action='store_true',
+        required=False,
+        help="if specified, perform more checks and emit more messages. "
+             "You may want to set TOAST_LOGLEVEL=DEBUG as well.",
+    )
+
+
+class SimpleCrosstalkMatrix:
+    """A thin crosstalk matrix class.
+
+    :param names: detector names.
+    :param data: crosstalk matrix.
+        The length of `names` should match the dimension of `data`,
+        which should be a square array.
+    :param debug: emit more debug message and perform runtime validation check.
+
+    This is a simple container storing the crosstalk matrix,
+    but not generating it.
+    """
+
+    def __init__(
+        self,
+        names: "Union[np.ndarray['S'], List[str]]",
+        data: 'np.ndarray[np.float64]',
+        *,
+        debug: 'bool' = False,
+    ):
+        self.names = np.asarray(names, dtype="S")
+        self.data = data
+        self.debug = debug
+        if debug:
+            self.__post_init__()
```

Comment:
This function is doing type / consistency checking.  This is important and not expensive compared to other operations, so why not always do that?

Reply:
resolved in commit 483c82e